source/opt/dead_branch_elim_pass.cpp

```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
+          uint32_t replId;
+          if (operands.size() == 2) {
+            replId = operands[0].words[0];
+          } else {
+            replId = TakeNextId();
+            std::unique_ptr<ir::Instruction> newPhi(
+                new ir::Instruction(context(), SpvOpPhi, inst->type_id(),
+                                    replId, std::move(operands)));
+            get_def_use_mgr()->AnalyzeInstDefUse(&*newPhi);
+            context()->set_instr_block(&*newPhi, &block);
+            inst->InsertBefore(std::move(newPhi));
           }
+          context()->ReplaceAllUsesWith(inst->result_id(), replId);
+          iter = context()->KillInst(&*inst);
+        } else {
+          ++iter;
         }
-        ++icnt;
-      });
-      // If just one live predecessor, replace resultid with live value id.
-      uint32_t replId;
-      if (lcnt == 1) {
-        replId = inst->GetSingleWordInOperand(lidx);
-      } else {
-        // Otherwise create new phi eliminating dead predecessor entries
-        assert(lcnt > 1);
-        replId = TakeNextId();
-        std::vector<ir::Operand> phi_in_opnds;
-        icnt = 0;
-        uint32_t lastId;
-        inst->ForEachInId(
-            [&deadPreds, &icnt, &phi_in_opnds, &lastId, this](uint32_t* idp) {
-              if (icnt % 2 == 1) {
-                if (deadPreds.find(cfg()->block(*idp)) == deadPreds.end()) {
-                  phi_in_opnds.push_back(
-                      {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {lastId}});
-                  phi_in_opnds.push_back(
-                      {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {*idp}});
-                }
-              } else {
-                lastId = *idp;
-              }
-              ++icnt;
-            });
-        std::unique_ptr<ir::Instruction> newPhi(new ir::Instruction(
-            context(), SpvOpPhi, inst->type_id(), replId, phi_in_opnds));
-        get_def_use_mgr()->AnalyzeInstDefUse(&*newPhi);
-        inst->InsertBefore(std::move(newPhi));
       }
-      const uint32_t phiId = inst->result_id();
-      context()->KillNamesAndDecorates(phiId);
-      (void)context()->ReplaceAllUsesWith(phiId, replId);
-      inst = context()->KillInst(inst);
     }
   }
 
-  // Erase dead blocks
-  for (auto ebi = func->begin(); ebi != func->end();)
-    if (elimBlocks.find(&*ebi) != elimBlocks.end())
+  // Erase dead blocks.
+  for (auto ebi = func->begin(); ebi != func->end();) {
+    if (unreachableMerges.count(&*ebi)) {
+      // Make unreachable, but leave the label.
+      KillAllInsts(&*ebi, false);
+      // Add unreachable terminator.
+      ebi->AddInstruction(
+          MakeUnique<ir::Instruction>(context(), SpvOpUnreachable, 0, 0,
+                                      std::initializer_list<ir::Operand>{}));
+      ++ebi;
+      modified = true;
+    } else if (unreachableContinues.count(&*ebi)) {
+      // Make unreachable, but leave the label.
+      KillAllInsts(&*ebi, false);
```

Comment:
This seems inconsistent with the handling of the Phi nodes above.  You delete the original branch, and all edges related to it, and add one back in for the header.  However, when updating the Phi node, you did not delete any edges.  You just changed it to Undef.  This is fine for the header, if the continue block branches to the header right away.  In other cases, the header will need a new entry for the Phi nodes.  For other block that are still live this will leave invalid Phi nodes.

I've assumed two things:

1) The back edge is not always in the continue block, only that it post-dominates the continue block.
2) The back edge can be a conditional branch, in the continue block, and the merge block can be reached some other way.

If these are both wrong, then there should not be a problem.

Reply:
If the continue is unreachable then blocks succeeding it on the backedge path are also unreachable. I guess what's missing from the phi update for the header is to add an undef to the phi from the continue instead which ever block it got a value from previously.

This is incorrect if OpLoopMerge does not imply structured control flow though (since I removed the requirement of Shader capability).